### PR TITLE
Added option to use DataViewGirds if Linux

### PR DESCRIPTION
**Pull request by ChrisMarsh82**
Added IGrid interface so that grids can be changed easily between DataViewGrid and Reo Grid. 
IGrid is used by the user control(ucrDataView, ucrColumnMetadata, and ucrDataframeMetadata) The ReoGrid and DataViewGrid implement the interface.

Issues within the DataViewGrid regarding editing the grid directly so have disabled for now

@dannyparsons having issues with setting SheetTabWidth, Can't work out why it keeps resetting.

### CODE REVIEW

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        End If
+    End Sub
+
+    Private Sub dataGrid_RowHeaderMouseClick(sender As Object, e As DataGridViewCellMouseEventArgs)
+        sender.SelectionMode = DataGridViewSelectionMode.RowHeaderSelect
+        If _rowContextMenuStrip IsNot Nothing And e.Button = MouseButtons.Right Then
+            If sender.SelectedRows.Count > 0 Then
+                _rowContextMenuStrip.Show(Cursor.Position)
+            End If
+        End If
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:244]: combine into single `if` (see above)?

**Comment by dannyparsons** [comment]
Thanks for this. Here's a few things I noticed:

**Windows Grid**
1. The sheets tab is not visible by default when I loaded a dataset. I had to reduce the size of the scroll bar to be able to see it.
2. The font seems to be different than before? Not a problem but probably safer to keep as it was.

image

3. When I had a filter, the info text was too long to fit in the Data View unless I make it bigger than half the screen. Can we reduce the text size or have it dynamically reduce?

image

4. Issue #6921 has reappeared. I think it just needs the forced refresh done in #6933 

Now going to look at the Linux grid...

File: instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        End Set
+    End Property
+
+    Public WriteOnly Property DataBook As clsDataBook Implements IGrid.DataBook
+        Set(value As clsDataBook)
+            _clsDataBook = value
+        End Set
+    End Property
+
+    Private Property IGrid_Enabled As Boolean Implements IGrid.Enabled
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb:41]:
```suggestion
    Private Property Grid_Enabled As Boolean Implements IGrid.Enabled
```

File: instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        Get
+            Return grdData.Enabled
+        End Get
+        Set(value As Boolean)
+            grdData.Enabled = value
+        End Set
+    End Property
+
+    Private Property IGrid_Visible As Boolean Implements IGrid.Visible
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb:50]:
```suggestion
    Private Property Grid_Visible As Boolean Implements IGrid.Visible
```

File: instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        Return lstSelectedRows
+    End Function
+
+    Public Function GetWorksheet(name As String) As clsWorksheetAdapter Implements IGrid.GetWorksheet
+        Dim worksheet = grdData.Worksheets.Where(Function(x) x.Name = name).FirstOrDefault
+        If worksheet IsNot Nothing Then
+            Return New clsWorksheetAdapter(worksheet)
+        Else
+            Return Nothing
+        End If
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb:93]:
```suggestion
        End If
        Return Nothing
```

File: instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    Public Function GetSelectedColumns() As List(Of clsColumnHeaderDisplay) Implements IDataViewGrid.GetSelectedColumns
+        Dim lstColumns As New List(Of clsColumnHeaderDisplay)
+        For i As Integer = grdData.CurrentWorksheet.SelectionRange.Col To grdData.CurrentWorksheet.SelectionRange.Col + grdData.CurrentWorksheet.SelectionRange.Cols - 1
+            lstColumns.Add(GetCurrentDataFrameFocus().clsVisiblePage.lstColumns(i))
+        Next
+        Return lstColumns
+    End Function
+
+    Public Function WorksheetCount() As Integer Implements IDataViewGrid.WorksheetCount
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb:78]:
```suggestion
    Public Function GetWorksheetCount() As Integer Implements IDataViewGrid.WorksheetCount
```

File: instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+'
+' You should have received a copy of the GNU General Public License
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Imports unvell.ReoGrid
+Imports unvell.ReoGrid.Events
+
+Public Class ucrDataframeMetadataReoGrid
+    Implements IDataframeMetaDataGrid
+    Private lstNonEditableColumns As New List(Of String)
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb:22]: where is this list populated?

File: instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    Protected Overrides Sub AttachEventsToWorksheet(worksheet As Worksheet)
+        AddHandler worksheet.AfterCellEdit, AddressOf Worksheet_AfterCellEdit
+        AddHandler worksheet.BeforeCellEdit, AddressOf Worksheet_BeforeCellEdit
+        AddHandler worksheet.BeforeCut, AddressOf Worksheet_BeforeCut
+        AddHandler worksheet.BeforePaste, AddressOf Worksheet_BeforePaste
+        AddHandler worksheet.BeforeRangeMove, AddressOf Worksheet_BeforeRangeMove
+        AddHandler worksheet.BeforeCellKeyDown, AddressOf Worksheet_BeforeCellKeyDown
+    End Sub
+
+    Private Sub Worksheet_AfterCellEdit(sender As Object, e As CellAfterEditEventArgs)
```
Comment (lloyddewit) [inline instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb:70]: remove underscores from this and following subs?

**Comment by dannyparsons** [comment]
I've looked at the Linux grid too. No further issues found. It's very nice that the sheet tabs are underneath the sheet scroll bar, I wish we had that in ReoGrid too!

**Comment by dannyparsons** [comment]
Note that 3. was a problem before this change.

File: instat/ucrColumnMetadata.vb [hunk 8]
```diff
@@ -46,303 +46,169 @@ Public Class ucrColumnMetadata
...
-        autoTranslate(Me)
-    End Sub
 
-    Private Sub grdVariables_CurrentWorksheetChange(sender As Object, e As EventArgs) Handles grdVariables.CurrentWorksheetChanged, Me.Load, grdVariables.WorksheetInserted
-        grdCurrSheet = grdVariables.CurrentWorksheet
-        grdCurrSheet.SetSettings(unvell.ReoGrid.WorksheetSettings.Edit_DragSelectionToMoveCells, False)
-        grdCurrSheet.SetSettings(unvell.ReoGrid.WorksheetSettings.Edit_DragSelectionToFillSerial, False)
-        grdCurrSheet.SelectionForwardDirection = unvell.ReoGrid.SelectionForwardDirection.Down
+        'DEBUG
+        ' If True Then
```
Comment (lloyddewit) [inline instat/ucrColumnMetadata.vb:97]: delete?

File: instat/ucrDataFrameMetadata.vb [hunk 9]
```diff
@@ -41,219 +45,143 @@ Public Class ucrDataFrameMetadata
...
-    ' TODO this needs tidying up
     Private Sub LoadForm()
         lstNonEditableColumns.AddRange({"class", "Is_Hidden", "Row_Count", "Column_Count", "Is_Linkable", "Is_Calculated"})
-        grdMetaData.Worksheets(0).Name = "metadata"
-        grdMetaData.SetSettings(unvell.ReoGrid.WorkbookSettings.View_ShowSheetTabControl, False)
-        grdMetaData.SetSettings(unvell.ReoGrid.WorkbookSettings.View_ShowHorScroll, True)
-        grdMetaData.SheetTabNewButtonVisible = False
+
+        'Debug
+        'If True Then
```
Comment (lloyddewit) [inline instat/ucrDataFrameMetadata.vb:52]: delete?

**Review by lloyddewit — CHANGES_REQUESTED** [review]

File: instat/ucrColumnMetadata.vb (hunk 8 above)
Comment (ChrisMarsh82) [reply instat/ucrColumnMetadata.vb:97]: This is here on purpose so we can quickly change between Reo and dataview grids.

File: instat/ucrDataFrameMetadata.vb (hunk 9 above)
Comment (ChrisMarsh82) [reply instat/ucrDataFrameMetadata.vb:52]: This is here on purpose so we can quickly change between Reo and dataview grids.

File: instat/Interface/IDataframeMetaDataGrid.vb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+Public Interface IDataframeMetaDataGrid
+    Inherits IGrid
+
+    Event EditValue(iRow As Integer, strColumnName As String, strPreviousValue As String, newValue As Object)
+
+    Sub AddColumns()
+
+    Sub AddRowData()
+
+    Sub SetMainWorksheetUp(strWorksheetName As String)
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Interface/IDataframeMetaDataGrid.vb:26]: This is confusing I will change it to SetupMainWorksheet. It did make sense if you read it in a Yorkshire accent.

File: instat/Interface/IDataViewGrid.vb [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+    Event WorksheetRemoved(worksheet As clsWorksheetAdapter)
+
+    Sub AddColumns(visiblePage As clsDataFramePage)
+
+    Sub AddRowData(dataFrame As clsDataFrame)
+
+    Function GetSelectedColumns() As List(Of clsColumnHeaderDisplay)
+
+    Function WorksheetCount() As Integer
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Interface/IDataViewGrid.vb:36]: Changed

File: instat/Interface/IGrid.vb [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    Property Visible() As Boolean
+    Function AddNewWorksheet(name As String) As clsWorksheetAdapter
+
+    Sub CopyRange()
+
+    Function GetCellValue(iRow As Integer, strColumn As String) As String
+
+    Function GetSelectedRows() As List(Of String)
+
+    Function GetWorksheet(name As String) As clsWorksheetAdapter
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Interface/IGrid.vb:32]: Changed

File: instat/Interface/IGrid.vb [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+' You should have received a copy of the GNU General Public License 
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Public Interface IGrid
+
+    Property CurrentWorksheet As clsWorksheetAdapter
+    WriteOnly Property DataBook As clsDataBook
+    WriteOnly Property InstatOptions As InstatOptions
+    Property Enabled() As Boolean
+    Property Visible() As Boolean
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Interface/IGrid.vb:23]: Changed

File: instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb (hunk 5 above)
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb:78]: Changed

File: instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb (hunk 7 above)
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/ReoGrid/ucrDataframeMetadataReoGrid.vb:70]: This is standard for Events. It is used in all the other events. i.e. controlName_Click

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+    Private Property IGrid_Enabled As Boolean Implements IGrid.Enabled
+        Get
+            Return tcTabs.Enabled
+        End Get
+        Set(value As Boolean)
+            tcTabs.Enabled = value
+        End Set
+    End Property
+
+    Private Property IGrid_Visible As Boolean Implements IGrid.Visible
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:58]: Auto generated

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+        End Set
+    End Property
+
+    Public WriteOnly Property InstatOptions As InstatOptions Implements IGrid.InstatOptions
+        Set(value As InstatOptions)
+            _InstatOptions = value
+        End Set
+    End Property
+
+    Public Function AddNewWorksheet(name As String) As clsWorksheetAdapter Implements IGrid.AddNewWorksheet
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:73]: Done

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+            End If
+        Next
+        Return ""
+    End Function
+
+    Public Function GetGrid(tabPage As TabPage) As DataGridView
+        Return CType(tabPage.Tag, DataGridView)
+    End Function
+
+    Public Function GetGrid(worksheetName As String) As DataGridView
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:110]: Done

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+        If _tabContextMenuStrip IsNot Nothing Then
+            AddHandler _tabContextMenuStrip.Opening, AddressOf contextMenuStrip_Opening
+            tcTabs.ContextMenuStrip = _tabContextMenuStrip
+        End If
+    End Sub
+
+    Public Sub SetCurrentDataFrame(strDataName As String) Implements IGrid.SetCurrentDataFrame
+        For Each tabPage As TabPage In tcTabs.TabPages
+            If tabPage.Text = strDataName Then
+                tcTabs.SelectedTab = tabPage
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:165]: Updated to exit loop

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+    End Function
+
+    Private Sub AttachContextMenuEventsToDataGrid(dataGrid As DataGridView)
+        AddHandler dataGrid.CellMouseClick, AddressOf dataGrid_CellMouseClick
+        AddHandler dataGrid.RowHeaderMouseClick, AddressOf dataGrid_RowHeaderMouseClick
+        AddHandler dataGrid.ColumnHeaderMouseClick, AddressOf dataGrid_ColumnHeaderMouseClick
+    End Sub
+
+    Private Sub contextMenuStrip_Opening(sender As Object, e As CancelEventArgs)
+        Dim p = tcTabs.PointToClient(Cursor.Position)
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb:210]: updated p and r

File: instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb (hunk 2 above)
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb:41]: Auto generated

File: instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb (hunk 3 above)
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/ReoGrid/ucrReoGrid.vb:50]: Auto generated

File: instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb (hunk 1 above)
Comment (ChrisMarsh82) [reply instat/UserControls/DataGrid/Linux/ucrLinuxGrid.vb]: done

**Review by dannyparsons — APPROVED** [review]
Tested and works well so merging.

**Pull request merged**